src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/TeragrepBloomFilter.java

```diff
@@ -54,117 +54,106 @@
 
     private final String partitionID;
-    private final byte[] bloomfilterBytes;
+    private final BloomFilter filter;
     private final Connection connection;
-    private final FilterSizes filterSizes;
-    private Long selectedExpectedNumOfItems;
-    private Double selectedFpp;
+    private final FilterTypes filterTypes;
 
-    public TeragrepBloomFilter(
-            String partition,
-            byte[] bloomfilterBytes,
-            Connection connection,
-            FilterSizes filterSizes
-    ) {
-        this.partitionID = partition;
-        this.bloomfilterBytes = bloomfilterBytes;
-        this.filterSizes = filterSizes;
-        this.connection = connection;
+    public TeragrepBloomFilter(String partition, byte[] bytes, Connection connection, FilterTypes filterTypes) {
+        this(partition, new ToBloomFilter(bytes), connection, filterTypes);
     }
 
-    private BloomFilter sizedFilter() {
-
-        SortedMap<Long, Double> filterSizesMap = filterSizes.asSortedMap();
-        Map<Long, Long> bitsizeToExpectedItemsMap = filterSizes.asBitsizeSortedMap();
-
-        try (ByteArrayInputStream bais = new ByteArrayInputStream(bloomfilterBytes)) {
-            BloomFilter bf = BloomFilter.readFrom(bais);
-
-            long bitSize = bf.bitSize();
-            if (bitsizeToExpectedItemsMap.containsKey(bitSize)) {
-                long expectedItems = bitsizeToExpectedItemsMap.get(bitSize);
-                double fpp = filterSizesMap.get(expectedItems);
-                this.selectedExpectedNumOfItems = expectedItems;
-                this.selectedFpp = fpp;
-                return bf;
-            }
-            else {
-                throw new IllegalArgumentException("no such filterSize <[" + bitSize + "]>");
-            }
-        }
-        catch (IOException e) {
-            throw new RuntimeException(e);
-        }
+    public TeragrepBloomFilter(String partition, BloomFilter filter, Connection connection, FilterTypes filterTypes) {
+        this.partitionID = partition;
+        this.filter = filter;
+        this.filterTypes = filterTypes;
+        this.connection = connection;
     }
 
     /**
      * Write filter bytes to database
-     * 
-     * @param overwriteExisting Set if existing filter data will be overwritten
+     *
+     * @param overwrite Set if existing filter data will be overwritten
      */
-    public void saveFilter(Boolean overwriteExisting) {
-
-        final BloomFilter filter = sizedFilter();
-        final String sql = sqlString(overwriteExisting);
-
-        try (PreparedStatement stmt = connection.prepareStatement(sql)) {
-            try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+    public void saveFilter(final Boolean overwrite) {
+        final long bitSize = filter.bitSize();
+        final long selectedExpectedNumOfItems;
+        final Double selectedFpp;
+        if (filterTypes.bitSizeMap().containsKey(bitSize)) {
```

Comment:
calling bitSizeMap() multiple times here parses json everytime, perhaps it would be beneficial to do it once and save to variable?

Reply:
made bitSizeMap into local variable